syntheseus/cli/search.py

```diff
@@ -165,14 +163,7 @@ def run_from_config(config: SearchConfig) -> Path:
         )
 
     # Load the single-step model
-    base_model = get_model(config, batch_size=1, num_gpus=int(config.use_gpu))  # type: ignore
-
-    # Set up the search algorithm
-    search_rxn_model = SyntheseusBackwardReactionModel(
-        cast(BackwardReactionModel, base_model),
-        num_results=config.num_top_results,
-        use_cache=config.reaction_model_use_cache,
-    )
+    search_rxn_model = get_model(config, batch_size=1, num_gpus=int(config.use_gpu))  # type: ignore
```

Comment:
Don't we still have to set `num_results` and `use_cache` here?

Reply:
That was unclear to me: since these are already in the config, are they not handled automatically by `get_model`? Maybe I am misunderstanding the internals of this method though.

Reply:
Yes, but `get_model` sees the `config` as a `BackwardModelConfig`, so hyperparameters added on top of that in `SearchConfig` are not accessed. Thus, `num_top_results` and `reaction_model_use_cache` are currently unused, and the defaults set there are ignored; running `search.py` out-of-the-box does produce your warnings about caching being turned off. For something to be passed to the single-step model it has to be captured in the `model_kwargs` argument to `BackwardModelConfig`, so e.g. caching can be turned on from CLI via:
```
model_kwargs="{'use_cache': True}"
```

I'm wondering what the best approach here is. We could in theory take `num_top_results` and `reaction_model_use_cache` and put those into `model_kwargs`, but that would lead to having two paths for specifying the same hyperparameter. Instead, we could delete those two arguments, assuming number of results and caching will be configured through `model_kwargs`, but we need to make sure the defaults are set right (currently they are not). Overriding defaults is tricky to marry with `SearchConfig` inheriting from `BackwardModelConfig`, but we could switch to having a field instead; something like
```python
class SearchConfig:
    model_config: BackwardModelConfig = BackwardModelConfig(
        model_kwargs={"use_cache": True, "default_num_results": 50}
    )
    ...
```

This would resolve the confusion and set the correct defaults. The only remaining thing would be that changing `default_num_results` sounds like a relatively common thing to explore, and setting it via CLI would be a bit clunky (also might lead to accidentally overriding the default for `use_cache`). If we want to improve this, we could modify `ModelConfig` to add `use_cache` and `default_num_results` explicitly (the `model_kwargs` bit was mostly intended for kwargs that are unique to a particular model class).

What do you think?

Reply:
Ah, I see the additional complication. None of these solutions seem ideal. What about just doing:

```python
search_rxn_model = get_model(config, batch_size=1, num_gpus=int(config.use_gpu))  # type: ignore
search_rxn_model._use_cache = config.reaction_model_use_cache
search_rxn_model.default_num_results = config.num_top_results
search_rxn_model.reset()
```

The only "bad" thing we are doing here is setting the private `_use_cache` attribute. However, since the model has not been called at this stage and we reset it immediately nothing unexpected should happen.

Reply:
Looks a bit hacky but also seems okay. A variant of this would be to add a `**kwargs` argument to `get_model` (and pass that directly to the model `__init__`), so that we can avoid changing things after construction:
```python
search_rxn_model = get_model(
    config,
    batch_size=1,
    num_gpus=int(config.use_gpu)),
    use_cache=config.reaction_model_use_cache,
    default_num_results=config.num_top_results,
)
```

Reply:
Good idea. I just pushed a commit with the following changes:

1. As you suggest, add **model_kwargs to get_model (with a check to ensure duplicate kwargs are not supplied)
2. Create the model with `get_model` as you suggested in your most recent comment, supplying these kwargs
3. Just in case we want it for the future, I added an option to change `use_cache` when calling `model.reset()`. This means that the "hacky" solution I suggested above could be done in a non-hacky way in the future.

What do you think of these changes?

Reply:
The newest changes look good to me!